### DISCUSSION TITLE
What is the logic behind "possible misuse of hexadecimal literal notation"?

### DISCUSSION THREAD

**Discussion by Greedquest** [twinbasic/documentation#1] [Q&A] [answered]
This code

```vba
i = ((l1 Xor &H8000) + l2 * 256) Xor &H8000
```

Gets warning:
```
 {WARNING} /vbInvoke/Sources/MemoryTools/TestLibMemory.bas [74,26]: Possible misuse of hexadecimal literal notation.  This value is interpretted as a signed Integer: -32768 (not 32768 as you might expect)
```

For me those Hex codes are clearly being used to set bits, rather than as numbers. So what is the logic there?

```vba
l = (((l1 + l2 * 256 + l3 * 256 ^ 2) Xor &H80000000) + l4 * 256 ^ 3) Xor &H80000000
```
this looks similar but there is no warning

**Comment by bclothier** [comment, marked as answer]
Those aren't the same thing. `&H8000` evaluates to -32768 whereas `&H8000&` evaluates to 32768. If the terms are all `Integer`s you won't notice any difference. However if they are `Long`, what you are doing is the difference between `&HFFFF8000&` vs `&H00008000&` which can mess up the bits in the upper portion of the value.

**Comment by KubaO** [comment]
Put simply
- `&H8000` is really `&H8000%`. Since the `%` type (integer) has range from -32768 to 32767, it can't represent 32768 (&H8000).
- `&H8000_0000` is really `&H8000_0000&`. Since the `&` type (long) has range -2_147_483_648 to 2_147_483_647, and it can't represent 2_147_483_648 (&H8000_0000) either. So there should be a warning there too.

> those Hex codes are clearly being used to set bits

Bits are numbers :)
